python/TestHarness/schedulers/Scheduler.py

```diff
@@ -151,23 +157,142 @@ def statusGo(self, tester_name):
+        # Local storage of downstream deleted tester objects
+        downstream_deleted = set([])
 
-        # Add a test to the Runner Queue, and ask a thread to do some work (the thread pool has built in methods
-        # which prevent oversubscribing)
+        # Local DAG
+        tester_dag = dag.DAG()
+
+        # Create the DAG, skipped_testers, and just about anything we can during this iteration loop
         for tester in testers:
-            results = self.runnerGo(tester.getTestName())
-            # DEBUG this thread by using the get method. (Note: this is a blocking call)
-            # print results.get()
+            with self.thread_lock:
+                # Create tester name to object map
+                self.scheduled_name_to_object[tester.getTestName()] = tester
+
+            # Create a node for this tester if not already exists (prereq iteration creates nodes)
+            if not tester_dag.node_exists(tester.getTestName()):
+                tester_dag.add_node(tester.getTestName())
+
+            # Loop through prereqs and add leaves/nodes where necessary
+            for prereq in tester.getPrereqs():
+
+                # Discover unknown prereqs
+                if prereq not in [x.getTestName() for x in testers]:
+                    tester.setStatus('unknown dependency', tester.bucket_fail)
+                    skipped_or_failed_tests.add(tester)
+                    continue
+
+                if not tester_dag.node_exists(prereq):
+                    # Create a node for this tester
+                    tester_dag.add_node(prereq)
+
+                # Once you add an edge to a node, the DAG arranges the nodes accordingly
+                try:
+                    tester_dag.add_edge(prereq, tester.getTestName())
+
+                # Handle Cyclic errors (tester status has already be applied)
+                except dag.DAGValidationError:
+                    pass
+
+            # Discover skipped tests
+            if not tester.getRunnable(self.options):
+                skipped_or_failed_tests.add(tester)
+
+        # Discover skipped tests and remove downstream tests that require this skipped test
+        for tester in skipped_or_failed_tests:
+            # Delete and return downstream nodes (prints 'skipped dependency' tests to the screen)
+            downstream_deleted = self.deleteDownstreamTests(tester_dag, tester)
+
+            # Delete this test
+            tester_dag.delete_node_if_exists(tester.getTestName())
+
+            # Inform the user it was skipped by adding it to the queue
+            self.statusGo(tester.getTestName())
+
+        # Get top level jobs we can launch
+        ind_jobs = tester_dag.ind_nodes()
+
+        # Maybe all jobs in this group got skipped?
+        if ind_jobs == []:
+            return
+
+        # Add our local DAG to the global shared dag
+        #with self.thread_lock:
+        self.shared_testers_dag[testers[-1].getTestDir()] = tester_dag
```

Comment:
I was fooling around with not using a lock. We probably should for this case. I've been reading its not necessary when creating a new node?